### PR TITLE
Feat: stateless block and transaction execution

### PR DESCRIPTION
**Pull request by mrLSD**
## Summary

This PR introduces **`aurora-evm-block-execution`**, a new, self-contained library that lifts the
single-transaction [`aurora-evm`](../evm) executor into a **stateless block-execution engine** built
for a zkEVM guest.

*Stateless* means the engine operates purely on an in-memory world-state snapshot — the kind that
would be reconstructed from a witness inside a proof — and never touches a database, the network, or
a mempool. A block is either validated **in full** or **rejected**, with no partially-mutated state
left observable.

Concretely, this PR lands the **transaction-execution core** of that engine (per-transaction
validation, exact gas/fee economics, typed receipts, gas & blob accounting) together with the
standalone, unit-tested **block primitives** it is built from (blob pricing & schedule, trie roots,
logs bloom, withdrawals, protocol requests, and the KZG point-evaluation precompile). It also
upgrades the whole workspace to **c-kzg 2.1**.

## Motivation

A zkEVM that validates Ethereum blocks needs an execution component that is (a) driven by a witness
rather than a live node, (b) fully deterministic so the same inputs always produce the same trace,
and (c) small and dependency-light so it is cheap to audit and to run inside a proof. The existing
`aurora-evm` core executes one transaction at a time and is not enough on its own. This crate fills
that gap without pulling in a full node stack or the `alloy` type ecosystem.

## What's in this PR

- A **new crate**, `aurora-evm-block-execution`, containing the transaction-execution engine and all
  block primitives (see the map below).
- A **workspace-wide upgrade to c-kzg 2.1** (used by the KZG point-evaluation precompile), with the
  test runner switched from a hand-parsed trusted setup to the mainnet setup bundled with c-kzg
  (this removes ~600 KB of vendored setup assets).
- Minor supporting adjustments to the `aurora-evm` core and the `evm-tests` runner required by the
  block layer, plus Rust 1.97 / edition-2024 lint conformance.

## The crate at a glance

| Module | Responsibility |
|---|---|
| `evm.rs` | The engine: `Evm` (owns block env, precompiles, world state), the transaction loop, per-transaction execution and fee settlement. |
| `evm_context.rs` | Per-transaction validation (header/field consistency, per-type checks, intrinsic + floor gas, required funds). |
| `blob.rs` | EIP-4844/7691/7918 blob pricing (`fake_exponential`, `calc_blob_gas_price`, `calc_excess_blob_gas`) and the EIP-7840/7892 timestamp-activated `BlobSchedule` / `BlobParams` (incl. the EIP-7594 per-transaction cap). |
| `precompiles/kzg.rs` | EIP-4844 point-evaluation precompile (`0x0a`) on c-kzg 2.1. |
| `transaction.rs` (+ `transaction/`) | The `Transaction` model, `TxType`, `TxKind`, `AccessList`, EIP-7825 gas cap. |
| `trie.rs` | `state_root` (two modes — see below), `receipts_root`, `withdrawals_root`, `transactions_root`. |
| `receipt.rs`, `bloom.rs` | Typed receipts and the 256-byte logs bloom. |
| `withdrawal.rs` | Validator withdrawals (EIP-4895). |
| `requests.rs` | Protocol requests hashing (EIP-7685; EIP-6110 deposits, EIP-7002 withdrawals, EIP-7251 consolidations). |
| `block.rs`, `spec.rs`, `result.rs`, `errors.rs` | Block environment / expected-header types, hardfork enum, result and typed error surfaces. |

### Public entry point

```rust
use aurora_evm_block_execution::evm::Evm;

// State is a witness-reconstructed snapshot; senders are already recovered upstream.
let evm = Evm::new(
    chain_id,        // Option<u64>
    block_env,       // header context, BLOCKHASH window, blob fee
    transactions,    // Vec<Transaction>, in block order
    spec,            // active hardfork
    state,           // BTreeMap<H160, MemoryAccount>
    &blob_schedule,  // validated at construction (see below)
)?;

let result = evm.execute_transactions()?;
// result.receipts / result.gas_used / result.blob_gas_used / result.state
```

## Gas economics

The per-transaction settlement mirrors, byte-for-byte, the proven reference already used by the
`evm-tests` state-test runner on the same core:

1. reserve `effective_gas_price * gas_limit` (plus the max blob fee) up front — **without** `value`,
   which the executor moves during the call;
2. run the call/create;
3. pay the coinbase its priority tip (from London the base fee is burned);
4. refund the caller its unused gas;
5. burn the base fee and the blob fee.

Cumulative gas and blob counts are accumulated with checked arithmetic, and a typed receipt is built
per transaction.

## Design principles

- **Concrete, single-path types.** One `struct Evm`, one `struct Transaction`, a closed `TxType`
  enum — no generics, traits, hooks, or executor factories. The crate targets a single
  Ethereum-mainline path, so there is nothing to abstract over.
- **Deterministic by construction.** Ordered `BTreeMap` state (reproducible roots and hashes),
  checked/saturating integer math with **no floating point**, and `#![forbid(unsafe_code)]`. Pricing
  helpers return `Option` and never panic or spin, even on adversarial witness input.
- **Dependency-light, `alloy`-free.** Types are defined on `primitive-types` + `rlp` + `triehash`;
  no external consensus/type ecosystem is required.
- **Spec-aligned and fail-closed.** Function names and formulas match the EIPs and the reference
  clients (e.g. `fake_exponential` is the EIP-4844 helper, kept verbatim for auditability); an
  unsupported configuration is rejected rather than silently degraded.

### Two ways to compute `state_root`

The world-state root depends on **all** accounts, not only the ones a block touches, so it is
handled in two modes: a pure `sec_trie_root` over a complete account map (for standalone use and
tests), and — when only a witness is available — emitting the post-state diff for an external
witness-backed sparse trie to root. All other roots are pure functions over complete lists and are
always available directly.

### Invalid configuration is unrepresentable

`BlobSchedule` can only be built through `try_new`, which sorts entries, rejects duplicate activation
timestamps, and validates every `BlobParams`; its inner list is private and its `Deserialize` routes
through the same constructor. Downstream pricing/limit code therefore never re-validates and can
assume well-formed parameters.

## Scope

**In this PR:** the transaction-execution engine (validation → execution → fee settlement →
receipts/accounting) and all the block primitives listed above, each unit-tested in isolation.

**Immediate follow-up:** a single `execute_block` entry point that wires the pieces together —
pre-execution system calls (EIP-2935 history, EIP-4788 beacon root), withdrawal crediting, request
gathering, root computation, and comparison against the expected header. The types, pure functions,
and error variants for these steps are already provided here; only the orchestration is deferred.

**Deliberately out of scope:** a full node (networking, mempool, payload building, consensus layer),
signature/sender recovery (senders arrive already recovered), reorg/revert machinery, and pre-merge
block/uncle rewards.

## Testing

- `make check` — `rustfmt` + `clippy` (`pedantic` + `nursery`, warnings denied) across the workspace: **clean**.
- `cargo test --all` — unit tests, including **79** in `aurora-evm-block-execution` covering blob math, schedule validation, transaction validation, gas economics, execution edge cases (revert, out-of-gas, create), and blob caps: **all passing**.
- Ethereum state tests via the `evm-tests` runner (GeneralStateTests + LegacyTests + stable fixtures): **187,090 passed, 0 failed** — this exercises the c-kzg 2.1 migration end-to-end, including the point-evaluation precompile against real KZG vectors.

## Review guide

Start with `evm.rs` (`Evm::new`, `run_transaction_loop`, `execute_validated`) for the engine and the
fee settlement, then `evm_context.rs` for per-transaction validation and `blob.rs` for blob pricing
and the schedule. The remaining modules (`trie.rs`, `receipt.rs`, `bloom.rs`, `withdrawal.rs`,
`requests.rs`, and the type modules) are self-contained and can be read independently.

## Summary by CodeRabbit

* **New Features**
  * Added a complete block execution pipeline supporting transaction validation, execution, receipts, state updates, and fork-specific block fields.
  * Added configurable Cancun, Prague, and Osaka blob parameters, schedules, pricing, limits, and EIP-7918 reserve-price handling.
  * Added support for withdrawals, beacon roots, requests, delegated accounts, and related fork validation.
* **Bug Fixes**
  * Improved transaction fee, gas, blob-data fee, and overflow handling.
  * Updated KZG proof verification and trusted-setup handling.
* **Documentation**
  * Clarified deterministic integer arithmetic principles for zkEVM compatibility.

### CODE REVIEW

**Comment by mrLSD** [comment]
@coderabbitai
Check and Learn:

the spec-typing observation is correct, but we'd like to push back on the **impact** (and therefore the severity), with evidence. Short version: `u128` does **not** reject any *valid* blob transaction, it matches what the Rust reference clients do, and our downstream fee math is already `U256`.

**1. The premise "u128 rejects otherwise-valid blob transactions" doesn't hold — affordability bounds the field far below `u128::MAX`.**
A blob transaction is only valid if the sender can afford it, which for the blob component means:

```
max_fee_per_blob_gas × total_blob_gas ≤ balance ≤ total_ETH_supply ≈ 1.2e26 wei ≈ 2^87
total_blob_gas ≥ GAS_PER_BLOB = 131072 = 2^17        (≥ 1 blob)
⇒ max_fee_per_blob_gas ≤ 2^87 / 2^17 = 2^70
```

`2^70 ≪ u128::MAX = 2^128` — `u128` represents every affordable (hence every potentially-valid) value with ~`2^58` (≈ 3×10¹⁷) of headroom. A value above `u128::MAX` would require a sender balance `> 2^145` wei ≈ 10²⁴ ETH, against a total supply of ~1.2×10⁸ ETH. Such a transaction is **unconditionally unaffordable → invalid**, so no valid transaction is lost.

**2. No consensus divergence.**
A blob tx with `max_fee_per_blob_gas > u128::MAX` is invalid, so any block containing it is invalid. Whether it is rejected at the type boundary or at the balance check, every conformant client rejects the same block. There is no input for which a `u128` client accepts a block a `uint256` client rejects (or vice versa), because the extra representable range is entirely unaffordable.

**3. This aligns us with the Rust reference clients — widening would diverge from them.**
- **revm**: `TxEnv.max_fee_per_blob_gas: u128` (`crates/context/src/tx.rs:78`; trait `fn max_fee_per_blob_gas(&self) -> u128` at `interface/src/transaction.rs:102`), and it promotes to `U256` only for the product (`transaction.rs:118`).
- **reth / alloy-consensus**: `max_fee_per_blob_gas` is `u128` throughout (`reth/crates/transaction-pool/src/traits.rs:1669`, `validate/mod.rs:372`; `alloy_consensus::TxEip4844.max_fee_per_blob_gas: u128`).
- Only **geth** carries 256-bit (`*uint256.Int`) — the natural type for Go's bigint, not a correctness requirement.

Our field type + the "promote to `U256` for the fee product" pattern is **identical to revm's**. Switching to `U256` would make us the outlier among Rust clients.

**4. Downstream fee arithmetic is already `U256` — there is no overflow risk to fix.**
`calc_max_data_fee` already does `U256::from(self.tx.max_fee_per_blob_gas).saturating_mul(U256::from(total_blob_gas))` (`evm_context.rs:103`); the only `u128`-domain use is the cap comparison `blob_gas_price > tx.max_fee_per_blob_gas` (`evm_context.rs:404`), where `blob_gas_price` is itself `u128` (and bounded the same way). So the "Heavy lift … downstream fee arithmetic" is largely already handled.

**5. Scope note.** This crate consumes **already-decoded** transactions (stateless; senders recovered upstream) — it does not RLP-decode raw tx bytes. Any narrowing to `u128` would happen in the upstream witness decoder, and again only for unaffordable values, so "rejects … before validation" does not describe this layer.

File: evm-block-execution/src/blob.rs [hunk 1]
```diff
@@ -50,3 +272,421 @@ pub fn get_total_blob_gas(blob_hashes_len: usize) -> u64 {
...
+        if execution_cost > blob_cost {
+            let delta = params
+                .max_blobs_per_block
+                .checked_sub(params.target_blobs_per_block)?;
+            let reserved = parent_blob_gas_used.checked_mul(delta)? / params.max_blobs_per_block;
+            return parent_excess_blob_gas.checked_add(reserved);
+        }
+    }
+
+    total.checked_sub(target_blob_gas)
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply evm-block-execution/src/blob.rs:367]: This ordering is intentional and matches the normative EIP-7918 pseudocode. The specification first returns zero when parent.excess_blob_gas + parent.blob_gas_used < target_blob_gas, and evaluates the reserve-price condition only afterward. Moving the reserve branch before this guard would produce a non-zero excess value where EIP-7918 requires zero and would itself cause consensus divergence.

[the remark this answers is not in the document]
Comment (mrLSD) [reply evm-block-execution/src/blob.rs:367]: @odin-by-borg
This ordering is intentional and matches the normative EIP-7918 pseudocode. The specification first returns zero when parent.excess_blob_gas + parent.blob_gas_used < target_blob_gas, and evaluates the reserve-price condition only afterward. Moving the reserve branch before this guard would produce a non-zero excess value where EIP-7918 requires zero and would itself cause consensus divergence.

**Comment by mrLSD** [comment]
@coderabbitai please provide full code review with security audit and relations to valid block-execution flow for zkEVM. Read carefully PR description and keep in mind discussions in PR.

**Comment by mrLSD** [comment]
@coderabbitai provide detailed audit
